pykokkos/lib/ufuncs.py

```diff
@@ -798,4 +798,538 @@ def logaddexp(viewA, viewB):
+    return out
+
+@pk.workunit
+def logical_or_impl_1d_double(tid: int, viewA: pk.View1D[pk.double], viewB: pk.View1D[pk.double], out: pk.View1D[pk.uint16]):
+    out[tid] = viewA[tid] or viewB[tid]
+
+@pk.workunit
+def logical_or_impl_1d_float(tid: int, viewA: pk.View1D[pk.float], viewB: pk.View1D[pk.float], out: pk.View1D[pk.uint16]):
+    out[tid] = viewA[tid] or viewB[tid]
+
+
+def logical_or(viewA, viewB):
+    """
+    Return the element-wise truth value of viewA OR viewB.
+
+    Parameters
+    ----------
+    viewA : pykokkos view
+            Input view.
+    viewB : pykokkos view
+            Input view.
+
+    Returns
+    -------
+    out : pykokkos view (uint16)
+           Output view.
+
+    """
+    out = pk.View([viewA.shape[0]], pk.uint16)
+    if str(viewA.dtype) == "DataType.double" and str(viewB.dtype) == "DataType.double":
+        pk.parallel_for(
+            viewA.shape[0],
+            logical_or_impl_1d_double,
+            viewA=viewA,
+            viewB=viewB,
+            out=out)
+
+    elif str(viewA.dtype) == "DataType.float" and str(viewB.dtype) == "DataType.float":
+        pk.parallel_for(
+            viewA.shape[0],
+            logical_or_impl_1d_float,
+            viewA=viewA,
+            viewB=viewB,
+            out=out)
+    else:
+        raise RuntimeError("Incompatible Types")
+    return out
+
+@pk.workunit
+def logical_xor_impl_1d_double(tid: int, viewA: pk.View1D[pk.double], viewB: pk.View1D[pk.double], out: pk.View1D[pk.uint16]):
+    out[tid] = bool(viewA[tid]) ^ bool(viewB[tid])
+
+@pk.workunit
+def logical_xor_impl_1d_float(tid: int, viewA: pk.View1D[pk.float], viewB: pk.View1D[pk.float], out: pk.View1D[pk.uint16]):
+    out[tid] = bool(viewA[tid]) ^ bool(viewB[tid])
+
+
+def logical_xor(viewA, viewB):
+    """
+    Return the element-wise truth value of viewA XOR viewB.
+
+    Parameters
+    ----------
+    viewA : pykokkos view
+            Input view.
+    viewB : pykokkos view
+            Input view.
+
+    Returns
+    -------
+    out : pykokkos view (uint16)
+           Output view.
+
+    """
+    out = pk.View([viewA.shape[0]], pk.uint16)
+    if str(viewA.dtype) == "DataType.double" and str(viewB.dtype) == "DataType.double":
+        pk.parallel_for(
+            viewA.shape[0],
+            logical_xor_impl_1d_double,
+            viewA=viewA,
+            viewB=viewB,
+            out=out)
+
+    elif str(viewA.dtype) == "DataType.float" and str(viewB.dtype) == "DataType.float":
+        pk.parallel_for(
+            viewA.shape[0],
+            logical_xor_impl_1d_float,
+            viewA=viewA,
+            viewB=viewB,
+            out=out)
+    else:
+        raise RuntimeError("Incompatible Types")
+    return out
+
+@pk.workunit
+def logical_not_impl_1d_double(tid: int, view: pk.View1D[pk.double], out: pk.View1D[pk.double]):
+    out[tid] = not view[tid]
+
+@pk.workunit
+def logical_not_impl_1d_float(tid: int, view: pk.View1D[pk.float], out: pk.View1D[pk.float]):
```

Comment:
Shouldn't `logical_not` have a return (`out`) type that is `bool`, which for us means `pk.uint8`?

Reply:
It depends how well we want to do it I guess. How many bits are used for the C++ bool? The `np.bool_` uses `8` bits I think, which may surprise some. Conversely, the type/`struct` may need to have a bit more metadata than just `1` or `0`, and I don't think Kokkos at C++ level uses `bool` directly in parallel loops?

Maybe we basically use `1` and `0` under the hood with a few "decorations" for printing them as `True` and `False` at the end of the day.

Reply:
Let's just have it as an alias to `int8` and then we can later change what is in background